sycl/test-e2e/syclcompat/launch/launch.cpp

```diff
@@ -0,0 +1,353 @@
+  syclcompat::launch<dynamic_local_mem_basicdt_kernel<T>>(ltt.range_3_,
+                                                          ltt.memsize_, d_a);
+  syclcompat::launch<dynamic_local_mem_basicdt_kernel<T>>(
+      ltt.grid_, ltt.thread_, ltt.memsize_, d_a);
+}
+
+template <typename T> void test_basic_dt_launch_q() {
+  std::cout << __PRETTY_FUNCTION__ << std::endl;
+
+  T d_a = T(1);
+  LaunchTestWithArgs<T> ltt;
+
+  if (ltt.skip) // Unsupported aspect
+    return;
+
+  syclcompat::launch<dynamic_local_mem_basicdt_kernel<T>>(
+      ltt.range_1_, ltt.memsize_, ltt.in_order_q_, d_a);
+  syclcompat::launch<dynamic_local_mem_basicdt_kernel<T>>(
+      ltt.range_2_, ltt.memsize_, ltt.in_order_q_, d_a);
+  syclcompat::launch<dynamic_local_mem_basicdt_kernel<T>>(
+      ltt.range_3_, ltt.memsize_, ltt.in_order_q_, d_a);
+  syclcompat::launch<dynamic_local_mem_basicdt_kernel<T>>(
+      ltt.grid_, ltt.thread_, ltt.memsize_, ltt.in_order_q_, d_a);
+}
+
+template <typename T> void test_arg_launch() {
+  std::cout << __PRETTY_FUNCTION__ << std::endl;
+
+  LaunchTestWithArgs<T> ltt;
+  if (ltt.skip) // Unsupported aspect
+    return;
+
+  T *d_a = (T *)syclcompat::malloc<T>(ltt.memsize_);
+
+  syclcompat::launch<dynamic_local_mem_typed_kernel<T>>(ltt.range_1_,
+                                                        ltt.memsize_, d_a);
+  syclcompat::launch<dynamic_local_mem_typed_kernel<T>>(ltt.range_2_,
+                                                        ltt.memsize_, d_a);
+  syclcompat::launch<dynamic_local_mem_typed_kernel<T>>(ltt.range_3_,
+                                                        ltt.memsize_, d_a);
+  syclcompat::launch<dynamic_local_mem_typed_kernel<T>>(ltt.grid_, ltt.thread_,
+                                                        ltt.memsize_, d_a);
+
+  syclcompat::free((void *)d_a);
+}
+
+template <typename T> void test_arg_launch_q() {
+  std::cout << __PRETTY_FUNCTION__ << std::endl;
+
+  LaunchTestWithArgs<T> ltt;
+  if (ltt.skip) // Unsupported aspect
+    return;
+
+  T *d_a = (T *)syclcompat::malloc<T>(ltt.memsize_, ltt.in_order_q_);
+
+  syclcompat::launch<dynamic_local_mem_typed_kernel<T>>(
+      ltt.range_1_, ltt.memsize_, ltt.in_order_q_, d_a);
+  syclcompat::launch<dynamic_local_mem_typed_kernel<T>>(
+      ltt.range_2_, ltt.memsize_, ltt.in_order_q_, d_a);
+  syclcompat::launch<dynamic_local_mem_typed_kernel<T>>(
+      ltt.range_3_, ltt.memsize_, ltt.in_order_q_, d_a);
+  syclcompat::launch<dynamic_local_mem_typed_kernel<T>>(
+      ltt.grid_, ltt.thread_, ltt.memsize_, ltt.in_order_q_, d_a);
+
+  syclcompat::free((void *)d_a, ltt.in_order_q_);
+}
+
+template <typename T> void test_local_mem_usage() {
+  std::cout << __PRETTY_FUNCTION__ << std::endl;
+
+  LaunchTestWithArgs<T> ltt;
+  if (ltt.skip) // Unsupported aspect
+    return;
+
+  size_t num_elements = ltt.memsize_ / sizeof(T);
+
+  T *h_a = (T *)syclcompat::malloc_host(ltt.memsize_);
+  T *d_a = (T *)syclcompat::malloc<T>(num_elements);
```

Comment:
Both `malloc` and `malloc_host` have templated & non-templated versions. I think we should be consistent & use either one or the other. If we use the templated version, omit the `(T *)` cast.

Same below for `test_local_mem_usage_q()`

Reply:
Every malloc was refactored to the non-templated version. So each call should be casted to `(T *)` and should allocate `ltt.memsize_` bytes.